### ISSUE TITLE
2.3.6-p1 breaks with PHP 7.2

### ISSUE THREAD

**Issue by kassner** [magento/magento2#32073]
### Preconditions

1. Magento 2.3.6;
2. PHP 7.2;

### Steps to reproduce

1. Upgrade Magento to 2.3.6-p1;

### Expected result

1. Store working;

### Actual result

1. _Any_ page in the store gives a 500 error and the exception log contains the following:

```
InvalidArgumentException: Failed to set ini option "session.cookie_samesite" to value "Lax". in /var/www/magento/vendor/magento/framework/Session/SessionManager.php:630
Stack trace:
#0 /var/www/magento/vendor/magento/framework/Session/SessionManager.php(197): Magento\Framework\Session\SessionManager->initIniOptions()
#1 /var/www/magento/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Framework\Session\SessionManager->start()
#2 /var/www/magento/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Framework\Session\Generic\Interceptor->___callParent('start', Array)
#3 /var/www/magento/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Framework\Session\Generic\Interceptor->Magento\Framework\Interception\{closure}()
#4 /var/www/magento/generated/code/Magento/Framework/Session/Generic/Interceptor.php(52): Magento\Framework\Session\Generic\Interceptor->___callPlugins('start', Array, Array)
#5 /var/www/magento/vendor/magento/framework/Session/SessionManager.php(141): Magento\Framework\Session\Generic\Interceptor->start()
#6 /var/www/magento/generated/code/Magento/Framework/Session/Generic/Interceptor.php(14): Magento\Framework\Session\SessionManager->__construct(Object(Magento\Framework\App\Request\Http), Object(Magento\Framework\Session\SidResolver\Proxy), Object(Magento\Framework\Session\Config), Object(Magento\Framework\Session\SaveHandler), Object(Magento\Framework\Session\Validator), Object(Magento\Framework\Session\Storage), Object(Magento\Framework\Stdlib\Cookie\PhpCookieManager), Object(Magento\Framework\Stdlib\Cookie\CookieMetadataFactory), Object(Magento\Framework\App\State\Interceptor), NULL)
#7 /var/www/magento/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(121): Magento\Framework\Session\Generic\Interceptor->__construct(Object(Magento\Framework\App\Request\Http), Object(Magento\Framework\Session\SidResolver\Proxy), Object(Magento\Framework\Session\Config), Object(Magento\Framework\Session\SaveHandler), Object(Magento\Framework\Session\Validator), Object(Magento\Framework\Session\Storage), Object(Magento\Framework\Stdlib\Cookie\PhpCookieManager), Object(Magento\Framework\Stdlib\Cookie\CookieMetadataFactory), Object(Magento\Framework\App\State\Interceptor), NULL)
#8 /var/www/magento/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php(66): Magento\Framework\ObjectManager\Factory\AbstractFactory->createObject('Magento\\Framewo...', Array)
#9 /var/www/magento/vendor/magento/framework/ObjectManager/ObjectManager.php(70): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create('Magento\\Framewo...')
#10 /var/www/magento/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(167): Magento\Framework\ObjectManager\ObjectManager->get('Magento\\Framewo...')
#11 /var/www/magento/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(273): Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgument(Array, 'Magento\\Framewo...', NULL, 'session', 'Magento\\Custome...')
#12 /var/www/magento/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(236): Magento\Framework\ObjectManager\Factory\AbstractFactory->getResolvedArgument('Magento\\Custome...', Array, Array)
#13 /var/www/magento/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php(34): Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgumentsInRuntime('Magento\\Custome...', Array, Array)
#14 /var/www/magento/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php(59): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->_resolveArguments('Magento\\Custome...', Array, Array)
#15 /var/www/magento/vendor/magento/framework/ObjectManager/ObjectManager.php(70): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create('Magento\\Custome...')
#16 /var/www/magento/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(167): Magento\Framework\ObjectManager\ObjectManager->get('Magento\\Custome...')
#17 /var/www/magento/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(273): Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgument(Array, 'Magento\\Custome...', NULL, 'customerSession', 'Magento\\Checkou...')
#18 /var/www/magento/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(236): Magento\Framework\ObjectManager\Factory\AbstractFactory->getResolvedArgument('Magento\\Checkou...', Array, Array)
#19 /var/www/magento/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php(34): Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgumentsInRuntime('Magento\\Checkou...', Array, Array)
#20 /var/www/magento/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php(59): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->_resolveArguments('Magento\\Checkou...', Array, Array)
#21 /var/www/magento/vendor/magento/framework/ObjectManager/ObjectManager.php(70): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create('Magento\\Checkou...')
#22 /var/www/magento/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(167): Magento\Framework\ObjectManager\ObjectManager->get('Magento\\Checkou...')
#23 /var/www/magento/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(273): Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgument(Array, 'Magento\\Checkou...', NULL, 'checkoutSession', 'Magento\\Persist...')
#24 /var/www/magento/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(236): Magento\Framework\ObjectManager\Factory\AbstractFactory->getResolvedArgument('Magento\\Persist...', Array, Array)
#25 /var/www/magento/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php(34): Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgumentsInRuntime('Magento\\Persist...', Array, Array)
#26 /var/www/magento/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php(59): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->_resolveArguments('Magento\\Persist...', Array, Array)
#27 /var/www/magento/vendor/magento/framework/ObjectManager/ObjectManager.php(70): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create('Magento\\Persist...')
#28 /var/www/magento/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(167): Magento\Framework\ObjectManager\ObjectManager->get('Magento\\Persist...')
#29 /var/www/magento/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(273): Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgument(Array, 'Magento\\Persist...', NULL, 'persistentSessi...', 'Magento\\Persist...')
#30 /var/www/magento/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(236): Magento\Framework\ObjectManager\Factory\AbstractFactory->getResolvedArgument('Magento\\Persist...', Array, Array)
#31 /var/www/magento/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php(34): Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgumentsInRuntime('Magento\\Persist...', Array, Array)
#32 /var/www/magento/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php(59): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->_resolveArguments('Magento\\Persist...', Array, Array)
#33 /var/www/magento/vendor/magento/framework/ObjectManager/ObjectManager.php(70): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create('Magento\\Persist...')
#34 /var/www/magento/vendor/magento/framework/Interception/PluginList/PluginList.php(250): Magento\Framework\ObjectManager\ObjectManager->get('Magento\\Persist...')
#35 /var/www/magento/vendor/magento/framework/Interception/Interceptor.php(119): Magento\Framework\Interception\PluginList\PluginList->getPlugin('Magento\\Framewo...', 'persistent_page...')
#36 /var/www/magento/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Framework\App\Http\Context\Interceptor->Magento\Framework\Interception\{closure}()
#37 /var/www/magento/generated/code/Magento/Framework/App/Http/Context/Interceptor.php(78): Magento\Framework\App\Http\Context\Interceptor->___callPlugins('getVaryString', Array, Array)
#38 /var/www/magento/vendor/magento/framework/App/Response/Http.php(108): Magento\Framework\App\Http\Context\Interceptor->getVaryString()
#39 /var/www/magento/generated/code/Magento/Framework/App/Response/Http/Interceptor.php(37): Magento\Framework\App\Response\Http->sendVary()
#40 /var/www/magento/vendor/magento/module-page-cache/Model/App/Response/HttpPlugin.php(25): Magento\Framework\App\Response\Http\Interceptor->sendVary()
#41 /var/www/magento/vendor/magento/framework/Interception/Interceptor.php(121): Magento\PageCache\Model\App\Response\HttpPlugin->beforeSendResponse(Object(Magento\Framework\App\Response\Http\Interceptor))
#42 /var/www/magento/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Framework\App\Response\Http\Interceptor->Magento\Framework\Interception\{closure}()
#43 /var/www/magento/generated/code/Magento/Framework/App/Response/Http/Interceptor.php(117): Magento\Framework\App\Response\Http\Interceptor->___callPlugins('sendResponse', Array, Array)
#44 /var/www/magento/vendor/magento/framework/App/ExceptionHandler.php(103): Magento\Framework\App\Response\Http\Interceptor->sendResponse()
#45 /var/www/magento/vendor/magento/framework/App/ExceptionHandler.php(70): Magento\Framework\App\ExceptionHandler->handleDeveloperMode(Object(Magento\Framework\App\Bootstrap), Object(InvalidArgumentException), Object(Magento\Framework\App\Response\Http\Interceptor))
#46 /var/www/magento/vendor/magento/framework/App/Http.php(156): Magento\Framework\App\ExceptionHandler->handle(Object(Magento\Framework\App\Bootstrap), Object(InvalidArgumentException), Object(Magento\Framework\App\Response\Http\Interceptor), Object(Magento\Framework\App\Request\Http))
#47 /var/www/magento/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Framework\App\Http->catchException(Object(Magento\Framework\App\Bootstrap), Object(InvalidArgumentException))
#48 /var/www/magento/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Framework\App\Http\Interceptor->___callParent('catchException', Array)
#49 /var/www/magento/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Framework\App\Http\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Bootstrap), Object(InvalidArgumentException))
#50 /var/www/magento/generated/code/Magento/Framework/App/Http/Interceptor.php(39): Magento\Framework\App\Http\Interceptor->___callPlugins('catchException', Array, Array)
#51 /var/www/magento/vendor/magento/framework/App/Bootstrap.php(267): Magento\Framework\App\Http\Interceptor->catchException(Object(Magento\Framework\App\Bootstrap), Object(InvalidArgumentException))
#52 /var/www/magento/pub/index.php(40): Magento\Framework\App\Bootstrap->run(Object(Magento\Framework\App\Http\Interceptor))
#53 {main}
```

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [x] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by kassner** [comment]
If you are stuck on PHP 7.2 until 2.3.7 comes around, here is a [patch](https://gist.github.com/kassner/07d648aa6ab132198340caa7f24580c1) for you.

Really shameful that the composer.json says PHP 7.1+ but feature-wise it breaks on <7.3. At least make the composer.json reflect what you are doing.

**Comment by hostep** [comment]
I've warned @fascinosum and @nathanjosiah about this 8 days ago on Slack that they would need to communicate properly to their users that 2.3.6-p1 would break on PHP 7.1 and 7.2. Unfortunately I haven't seen anything mentioned in the security bulletins around this 🤷‍♂️ 
They did confirm they would change the composer constraints in Magento 2.3.7 to prevent this from happening in the future, but that's still a few months away.

I don't think that patch is a good idea, it's probably better to upgrade to PHP 7.3 if you can.

**Comment by kassner** [comment]
If you run Ubuntu LTS, you either patch it or keep the store offline until 2.3.7 is available.

**Comment by hostep** [comment]
But Magento 2.3.7 isn't going to be able to run on PHP 7.2 either if I understand everything correctly ...

**Comment by kassner** [comment]
18.04 has PHP 7.2, 20.04 has PHP 7.4. There is no version with PHP 7.3.

**Comment by kassner** [comment]
Anyway, that's not the issue. The issue is that the composer.json does not reflect the project dependencies.

**Comment by pemann** [comment]
So to patch this we should override Magento\Backend\Model\Auth\Session in adminhtml/di.xml and Magento\Framework\Session\Generic in frontend/di.xml ? Anything else?

**Comment by pemann** [comment]
> So to patch this we should override Magento\Backend\Model\Auth\Session in adminhtml/di.xml and Magento\Framework\Session\Generic in frontend/di.xml ? Anything else?

Actually it is better to do Auth\Session override in etc/di.xml so it will work both admin & rest.

**Comment by Calypso971** [comment]
Ran into this too (Ubuntu 18.04LTS, so bound to php7.2. Wanted to start upgrade the testenvironment to 20.04 some time ago and that isn't possible because M2.3 doesn't support php7.4 yet. Upgrading to M2.4 is considered more tricky at this moment because of customisations and used theme - testing has to be very extensive).

To be honest, I think this is a disgrace for Magento; just apparantly dropping support for PHP versions in a patch upgrade. Not even a minor version, but a patch upgrade.

I'm not a fan of "fixes" like proposed in this topic, but to be honest: running on unpatched Magento versions with known (security) bugs in it give me even more creeps...

**Comment by nathanjosiah** [comment]
PHP 7.2 has reached end of life, To maintain PCI compliance, Magento should not be run on unsupported software. Additionally our docs have been updated that Magento supports PHP 7.3
https://devdocs.magento.com/guides/v2.3/install-gde/system-requirements.html#php

> Magento supports PHP 7.3.
> 
> PHP 7.1 has reached End of Life. To maintain PCI compliance, Magento should not be run on unsupported software.
(which now would also include 7.2)

However after some discussion we have opened bugs to remove the composer support and we are also going to update our docs to show we no longer support 7.1 and 7.2. We are also going to mention the known issue that composer still shows support for 7.1 and 7.2.

**Comment by pemann** [comment]
Funny, how you tried to built the support for 7.2 but the very first naive test (load any page) showed that it did not work.
https://github.com/magento/magento2/blob/2.3.6-p1/lib/internal/Magento/Framework/Stdlib/Cookie/PhpCookieManager.php#L333

Is there anything else important in 2.3.6-p1 besides Lax-cookies?

**Comment by nathanjosiah** [comment]
@pemann I don't understand what you are saying. I explicitly said we do Not support 7.2 because it is not PCI compliant. It may not load but it isn't supported and isn't tested.

As for the value of 2.3.6-p1 it is a security release plus major required hotfixes. The security details can be found on https://helpx.adobe.com/security/products/magento/apsb21-08.html

**Comment by Calypso971** [comment]
> PHP 7.2 has reached end of life, To maintain PCI compliance, Magento should not be run on unsupported software. Additionally our docs have been updated that Magento supports PHP 7.3

I understand what you're saying, but Magento 2.3 users on Ubuntu (and possibly deriviates as Mint?) are screwed; they cannot upgrade PHP to 7.4 because you only start supporting it under 2.3.7, and that is not released yet, so you "need" to stay at 2.3.6 knowing that there are security fixes in 2.3.6p1... great...

My conclusion: PHP7.4 support should already have been in 2.3.6 or even 2.3.6p1. Because Ubuntu server's share isn't a marginal one.

**Comment by pemann** [comment]
> @pemann I don't understand what you are saying. I explicitly said we do Not support 7.2 because it is not PCI compliant. It may not load but it isn't supported and isn't tested.
> 
> As for the value of 2.3.6-p1 it is a security release plus major required hotfixes. The security details can be found on https://helpx.adobe.com/security/products/magento/apsb21-08.html

Someone wrote the code to support 7.2 (and older) but it seems that he/she did not run / test it on 7.2 even once. That is pretty optimistic development. 

Thanks for the link! It seems that patching is the way to go, since there are some other critical fixes.

**Comment by Calypso971** [comment]
> 
> 
> Funny, how you tried to built the support for 7.2 but the very first naive test (load any page) showed that it did not work.
> https://github.com/magento/magento2/blob/2.3.6-p1/lib/internal/Magento/Framework/Stdlib/Cookie/PhpCookieManager.php#L333
> 
> Is there anything else important in 2.3.6-p1 besides Lax-cookies?

testing isn't conclusive here (install was yesterday, first test is only to visit the testsite and do a testorder), but with the "not-official-patch" that test worked. Maybe we're encountering other strange things in the tests today/tomorrow, but what you are saying that no page loads isn't my experience.

**Comment by kassner** [comment]
Usually the GitHub issues stay open until a fix is made public 🙄 

Heck, the `2.3` branch here on GitHub doesn't even have the 2.3.6-p1 patch in it :|

> I explicitly said we do Not support 7.2 because it is not PCI compliant

@nathanjosiah that's not true. PHP 7.2 is PCI compliant if you are using Ubuntu LTS. Just because the _PHP Group_ isn't supporting PHP 7.2 does not mean no one else is providing that support.

**Comment by pemann** [comment]
@magento give me 2.4-develop instance

**Comment by nathanjosiah** [comment]
@Calypso971 
> but Magento 2.3 users on Ubuntu (and possibly deriviates as Mint?) are screwed; they cannot upgrade PHP to 7.4 because you only start supporting it under 2.3.7, and that is not released yet, so you "need" to stay at 2.3.6 knowing that there are security fixes in 2.3.6p1... great...

I understand but PHP 7.3 is still available and still supported.

@kassner 
The release is available https://github.com/magento/magento2/releases/tag/2.3.6-p1

>Just because the PHP Group isn't supporting PHP 7.2 does not mean no one else is providing that support.
This may be true but this is not official support and we cannot be PCI compliant and support a software that is officially EOL.

**Comment by Calypso971** [comment]
> > but Magento 2.3 users on Ubuntu (and possibly deriviates as Mint?) are screwed; they cannot upgrade PHP to 7.4 because you only start supporting it under 2.3.7, and that is not released yet, so you "need" to stay at 2.3.6 knowing that there are security fixes in 2.3.6p1... great...
> 
> I understand but PHP 7.3 is still available and still supported.

You are clearly not standing into the reality of Linux distributions here: they don't carry every PHP version. In Ubuntu the current LTS versions 7.2 and 7.4 are available. The only way installing 7.3 in it is through 3rd party repositories or by hand - both are not what you want in a professional environment: it can haunt you for quite some time. 

But the real problem here is that Magento in the 2.3 branch (if you want to stay up to date ofcourse, but who wants a webstore with known security issues) now only supports 1 single PHP version. Support for 7.4 (released in November 2019) will only come in the next release, nearing the 2 year mark after the release of PHP 7.4... Saying to customers who are running 7.2 things like "it's not supported anymore, you should have known that and be prepared" goes the same or even more for you as a developer: 7.4 is there for quite some time now, why isn't support for it available yet?!

Support for 7.2 dropped 30 November 2020; M2.3.6 was released on 15th of October 2020. If PCI compliance was so important, you could already have dropped 7.2 support then. That at least was not a minor patchrelease. In normal software terms: a patchrelease as 2.3.6-p1 should only patch and definitely not break functionality or change requirements. This 2.3.6-p1 does. 

But I understand your message: customer, you're screwed, deal with it yourself.
